File: app/src/main/java/com/yeolabgt/mahmoodms/androidssvepinterfacetf/GraphAdapter.java
package com.yeolabgt.mahmoodms.androidssvepinterfacetf;

import com.androidplot.xy.LineAndPointFormatter;
import com.androidplot.xy.SimpleXYSeries;

/**
 * Created by Musa Mahmood on 5/15/2017.
 * This file controls the AndroidPlot graph.
 */

class GraphAdapter {
    // Variables
    SimpleXYSeries series;
    LineAndPointFormatter lineAndPointFormatter;
    private int seriesHistoryDataPoints;
    int sampleRate;
    private double xAxisIncrement;
    boolean plotData;

    // Set/Get Methods (Don't need yet)
    void setPlotData(boolean plotData) {
        this.plotData = plotData;
        if (!plotData) {
            clearPlot();
        }
    }

    // Constructor
    GraphAdapter(int seriesHistoryDataPoints, String XYSeriesTitle, boolean useImplicitXVals, int lineAndPointFormatterColor) {
        //default values
        this.seriesHistoryDataPoints = seriesHistoryDataPoints;
        this.lineAndPointFormatter = new LineAndPointFormatter(lineAndPointFormatterColor, null, null, null);
        setPointWidth(5); //Def value:
        this.series = new SimpleXYSeries(XYSeriesTitle);
        if(useImplicitXVals) this.series.useImplicitXVals();
        this.plotData = false; //Don't plot data until explicitly told to do so:
    }

    void setPointWidth(float width) {
        this.lineAndPointFormatter.getLinePaint().setStrokeWidth(width);
    }

    void setSeriesHistoryDataPoints(int seriesHistoryDataPoints) {
        this.seriesHistoryDataPoints = seriesHistoryDataPoints;
    }

    void addDataPointsGeneric(double[] xdata, double[] ydata, int istart, int iend) {
        if(this.plotData) {
            for (int i = istart; i < iend; i++) {
                plot(xdata[i],ydata[i]);
            }
        }
    }

    void addDataPointTimeDomainAlt(double data, int index) {
        if(this.plotData) plot((double)index*xAxisIncrement,data);
    }

    void addDataPointTimeDomain(double data, int index) {
        if(this.plotData) plot((double)index*xAxisIncrement,data);
    }

    void setxAxisIncrementFromSampleRate(int sampleRate) {
        this.sampleRate = sampleRate;
        setxAxisIncrement((double)1/(double)sampleRate);
    }

    private void setxAxisIncrement(double xAxisIncrement) {
        this.xAxisIncrement = xAxisIncrement;
    }

    //Graph Stuff:
    void clearPlot() {
        if(this.series!=null) {
            while(this.series.size()>0) {
                this.series.removeFirst();
            }
        }
    }

    private void plot(double x, double y) {
        while (series.size()>seriesHistoryDataPoints-1) {
            series.removeFirst();
        }
        series.addLast(x,y);
    }

    private void plotAlt(double x, double y) {
        while (series.size()>seriesHistoryDataPoints-1) {
            series.removeLast();
        }
        series.addLast(x,y);
    }
}

File: app/src/main/java/com/yeolabgt/mahmoodms/ssvepinterfacetf/GraphAdapter.kt
package com.yeolabgt.mahmoodms.ssvepinterfacetf

import com.androidplot.xy.LineAndPointFormatter
import com.androidplot.xy.SimpleXYSeries

/**
 * Created by Musa Mahmood on 5/15/2017.
 * This file controls the AndroidPlot graph.
 */

internal class GraphAdapter
(private var seriesHistoryDataPoints: Int, XYSeriesTitle: String, useImplicitXVals: Boolean, lineAndPointFormatterColor: Int) {
    // Variables
    var series: SimpleXYSeries? = null
    var lineAndPointFormatter: LineAndPointFormatter = LineAndPointFormatter(lineAndPointFormatterColor, null, null, null)
    var sampleRate: Int = 0
    private var xAxisIncrement: Double = 0.toDouble()
    var plotData: Boolean = false

    init {
        setPointWidth(5f) //Def value:
        this.series = SimpleXYSeries(XYSeriesTitle)
        if (useImplicitXVals) this.series!!.useImplicitXVals()
        this.plotData = false //Don't plot data until explicitly told to do so:
    }

    fun setPointWidth(width: Float) {
        this.lineAndPointFormatter.linePaint.strokeWidth = width
    }

    fun setSeriesHistoryDataPoints(seriesHistoryDataPoints: Int) {
        this.seriesHistoryDataPoints = seriesHistoryDataPoints
    }

    fun addDataPointsGeneric(xdata: DoubleArray, ydata: FloatArray, istart: Int, iend: Int) {
        if (this.plotData) {
            for (i in istart until iend) {
                plot(xdata[i], ydata[i])
            }
        }
    }

    fun addDataPointTimeDomain(data: Double, index: Int) {
        if (this.plotData) plot(index.toDouble() * xAxisIncrement, data)
    }

    fun addDataPointTimeDomainAlt(data: Double, index: Int) {
        if (this.plotData) plotAlt(index.toDouble() * xAxisIncrement, data)
    }

    fun setxAxisIncrementFromSampleRate(sampleRate: Int) {
        this.sampleRate = sampleRate
        setxAxisIncrement(1.toDouble() / sampleRate.toDouble())
    }

    private fun setxAxisIncrement(xAxisIncrement: Double) {
        this.xAxisIncrement = xAxisIncrement
    }

    //Graph Stuff:
    fun clearPlot() {
        if (this.series != null) {
            while (this.series!!.size() > 0) {
                this.series!!.removeFirst()
            }
        }
    }

    private fun plot(x: Double, y: Float) {
        series!!.addLast(x, y)
    }

    private fun plot(x: Double, y: Double) {
        while (series!!.size() > seriesHistoryDataPoints - 1) {
            series!!.removeFirst()
        }
        series!!.addLast(x, y)
    }

    private fun plotAlt(x: Double, y: Double) {
        while (series!!.size() > seriesHistoryDataPoints - 1) {
            series!!.removeLast()
        }
        series!!.addLast(x, y)
    }
}
